starlite/types/parsed_signature.py

```diff
@@ -75,7 +75,60 @@ def is_optional(self) -> bool:
     @property
     def is_collection(self) -> bool:
         """Whether the annotation is a collection type or not."""
-        return self.origin and issubclass(self.origin, Collection)
+        return bool(self.origin and self.origin is not Union and issubclass(self.origin, Collection))
+
+    def is_type_of(self, cl: type[Any]) -> bool:
+        """Whether the annotation is a subclass of the given type.
+
+        Args:
+            cl: The type to check.
+
+        Returns:
+            Whether the annotation is a subclass of the given type.
+        """
+        if self.origin:
+            if self.origin is Union:
+                return any(t.is_type_of(cl) for t in self.inner_types)
```

Comment:
```py
    assert ParsedType.from_annotation(Union[int, str]).is_type_of(int) is True
    assert ParsedType.from_annotation(Union[int, str]).is_type_of(str) is True
    assert ParsedType.from_annotation(Union[int, str]).is_type_of(bool) is False
    assert ParsedType.from_annotation(Optional[int]).is_type_of(int) is True
    assert ParsedType.from_annotation(Optional[int]).is_type_of(str) is False
```
@starlite-api/maintainers is this the right logic for unions?

Reply:
Hm. I wouldn't say it's wrong but maybe unexpected?

How about making `is_type_of` accept `cl: type[Any] | tuple[type[Any], ...]`, and if it's a tuple, check that it matches all items (if it's a union / generic alias), and if not a tuple behave like it does now?
The current behaviour is what I'd expect from `has_type_of` when operating on a union and a single type.

Then you'd have

```python
assert ParsedType.from_annotation(Union[int, str]).is_type_of(int) is False
assert ParsedType.from_annotation(Union[int, str]).is_type_of((int, str)) is True
assert ParsedType.from_annotation(Union[int, str]).has_type_of(str) is True
assert ParsedType.from_annotation(Union[int, str]).has_type_of(int) is True
```

Reply:
Yeh, it isn't right.

Actually, I think if the types are a tuple it should behave like a tuple 2nd arg to `is_subclass`.

What would you say to just making this return `False` for unions?

We could consider adding another predicate property, such as `ParsedType.is_union`, and then explicit union introspection would look something like `if parsed_type.is_union and parsed_type.has_type_of(Thing)`. That is probably the right amount of information to leave behind for future readers of code that happens to be calling out to these things. As it stands, it would be unclear that the underlying call was handling the union aspect of the type check, which would probably make calling code harder  to understand.

I'd say:

```py
assert ParsedType.from_annotation(Union[int, str]).is_type_of((int, str)) is False
assert ParsedType.from_annotation(Union[int, str]).is_type_of(Union[int, str]) is True
# and probably
assert ParsedType.from_annotation(Union[int, str]).is_type_of(Union[int, str, None) is True
```
Should this be named `is_subtype_of()`?

Reply:
The issue with `if parsed_type.is_union and parsed_type.has_type_of(Thing)` is that it only works for unions; If I want to compare to other generic aliases (e.g. `dict[str, str]`) )`is_union` this approach does not work.

How would I then check if e.g. an annotation is strictly equal to `dict[str, str]` (ignoring `__future__` and different typing versions of course)?

> Should this be named `is_subtype_of()`?

That would certainly be an option. And make the intent clearer.

Reply:
> How would I then check if e.g. an annotation is strictly equal to dict[str, str] (ignoring __future__ and different typing versions of course)?

It would be `parsed_type.origin is dict and all(t.is_type_of(str) for t in parsed_type.inner_types)`.

> That would certainly be an option. And make the intent clearer.

Yeh, I'll rename.

Reply:
I think I'll just return `False` for `is_subtype_of()` union case for now, and we can add some handling for the unions when we have a concrete case to look at.

Reply:
> It would be `parsed_type.origin is dict and all(t.is_type_of(str) for t in parsed_type.inner_types)`.

Quite a mouthful :grimacing: And also not that great for error messages I think.

But let's leave that as is now. I have a concrete use case for this, but let's get this one merge for now and then I'll play around with it and see what works best in that case.

Reply:
> Quite a mouthful

Maybe something like `parsed_type.has_origin_and_types(dict, (str, str))`

> I have a concrete use case for this

Awesome, happy for you to work out what is best in that case.

I'll get this polished up.

Reply:
> Maybe something like `parsed_type.has_origin_and_types(dict, (str, str))`

At that point, why not just make `parsed_type == dict[str, str]` or `parsed_type == ParseType(origin=dict, args=(str, str))` work? For strict equality checking this might be the most intuitive solution (this is also how `GenericAlias` does equality comparison).

Reply:
> `parsed_type == dict[str, str]`

Oh that's genius! If that is effectively an alias for `parsed_type.has_origin_and_types(dict, (str, str))` that would be an awesome API.

```py
class ParsedType:
    def __eq__(self, annotation):
        other = ParsedType.from_annotation(annotation)
        return self.has_origin_and_types(other.origin, tuple(t.annotation for t in other.inner_types))
```
Something like that? It could still be useful to have the `has_origin_and_types()` method for dynamic use when the args aren't known, to save constructing the dynamic generic type at runtime.

> `parsed_type == ParseType(origin=dict, args=(str, str))`

I'm not so warm on this as I don't want `ParsedType` to be easy to instantiate and would rather it is always done through the `from_annotation()` method so we have the consistency that brings.

Reply:
> I'm not so warm on this as I don't want ParsedType to be easy to instantiate and would rather it is always done through the from_annotation() method so we have the consistency that brings.

I could always use `__post_init__()` too, and make the `__init__()` method only receive the annotation.

Reply:
> how do we handle Python 3.8 and 3.9? You can't do dict[str, str] there outside annotations, so ParsedType.from_annotation(dict[str, str]) would just be type error.

`ParsedType.from_annotation(Dict[str, str])` - same as always on those py versions.

> or can you think of a different way that supports the case of comparing against types from a constant that are not an annotation?

I don't understand what you mean - what is the constant, how is it different from a resolved type annotation?

Reply:
Oh, are you are saying that you want to write code that supports `parsed_type == dict[str, str]` and have it work on 3.8? Obviously that isn't  possible, but why can't you do `parsed_annotation == Dict[str, str]` in our code - it won't matter if the user has defined their annotation `dict[str, str]`.

Reply:
I think we are beyond the scope of this PR so I'll resolve this conversation so I can merge, however happy to keep talking about this - I'm obviously missing some understanding of the use-case, so would be good to get to the bottom of it.